core/dataset.cpp

```diff
@@ -680,4 +755,192 @@ std::ostream &operator<<(std::ostream &os, const Dim dim) {
+  Dataset res(lhs);
+  apply_with_delay(plus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator+(Dataset &&lhs, const Dataset &rhs) {
+  return apply(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(Dataset &&lhs, const DatasetConstProxy &rhs) {
+  return apply(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(Dataset &&lhs, const DataConstProxy &rhs) {
+  return apply_with_delay(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(const Dataset &lhs, Dataset &&rhs) {
+  return apply(plus_equals, rhs, lhs);
+}
+
+Dataset operator+(Dataset &&lhs, Dataset &&rhs) {
+  return apply(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(const DatasetConstProxy &lhs, const Dataset &rhs) {
+  Dataset res(lhs.dataset());
+  apply(plus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator+(const DatasetConstProxy &lhs, const DatasetConstProxy &rhs) {
+  Dataset res(lhs.dataset());
+  apply(plus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator-(const Dataset &lhs, const Dataset &rhs) {
+  Dataset res(lhs);
+  apply(minus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator-(const Dataset &lhs, const DatasetConstProxy &rhs) {
+  Dataset res(lhs);
+  apply(minus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator-(const Dataset &lhs, const DataConstProxy &rhs) {
+  Dataset res(lhs);
+  apply_with_delay(minus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator-(Dataset &&lhs, const Dataset &rhs) {
+  return apply(minus_equals, lhs, rhs);
+}
+
+Dataset operator-(Dataset &&lhs, const DatasetConstProxy &rhs) {
+  return apply(minus_equals, lhs, rhs);
+}
+
+Dataset operator-(Dataset &&lhs, const DataConstProxy &rhs) {
+  return apply_with_delay(minus_equals, lhs, rhs);
+}
+
+Dataset operator-(Dataset &&lhs, Dataset &&rhs) {
+  return apply(minus_equals, lhs, rhs);
+}
+
+Dataset operator-(const DatasetConstProxy &lhs, const Dataset &rhs) {
+  Dataset res(lhs.dataset());
+  apply(minus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator-(const DatasetConstProxy &lhs, const DatasetConstProxy &rhs) {
+  Dataset res(lhs.dataset());
+  apply(minus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator*(const Dataset &lhs, const Dataset &rhs) {
+  Dataset res(lhs);
+  apply(times_equals, res, rhs);
+  return res;
+}
+
+Dataset operator*(const Dataset &lhs, const DatasetConstProxy &rhs) {
+  Dataset res(lhs);
+  apply(times_equals, res, rhs);
+  return res;
+}
+
+Dataset operator*(const Dataset &lhs, const DataConstProxy &rhs) {
+  Dataset res(lhs);
+  apply_with_delay(times_equals, res, rhs);
+  return res;
+}
+
+Dataset operator*(Dataset &&lhs, const Dataset &rhs) {
+  return apply(times_equals, lhs, rhs);
+}
+
+Dataset operator*(Dataset &&lhs, const DatasetConstProxy &rhs) {
+  return apply(times_equals, lhs, rhs);
+}
+
+Dataset operator*(Dataset &&lhs, const DataConstProxy &rhs) {
+  return apply_with_delay(times_equals, lhs, rhs);
+}
+
+Dataset operator*(Dataset &&lhs, Dataset &&rhs) {
+  return apply(times_equals, lhs, rhs);
```

Comment:
I do not quite understand this and other similar other overloads. What is the point of passing both arguments by rvalue reference? And even if we have just one rvalue ref, does it not need to be moved to have any effect? Furthermore, `apply` accepts arguments by reference (as far I I can see), so even if we move here, we would still need to adapt `apply`?

Reply:
Overloads where both arguments are rvalue references are needed in the case when you have both rvalue - lvalue and lvalue - rvalue pairs, which are ambiguous for rvalue - ravlue.

This is only actually needed for `operator+` however.

Reply:
But as long as you have no code that determines which one to move from I do not see the point. Just use const lvalue ref then?

Reply:
Yes, I've removed it for all but `operator+` where it is required.

Reply:
Can you elaborate, why is it required for `+`?

Reply:
Cause `operator+` has (among others) the overloads:
```cpp
Dataset &&operator+(Dataset &&lhs, const Dataset &rhs);
Dataset &&operator+(const Dataset &lhs, Dataset &&rhs);
```

Without `Dataset &&operator+(Dataset &&lhs, Dataset &&rhs);` this operator is ambiguous when both sides are rvalues.

Reply:
Why only for `+`?

In any case, didn't we conclude in a slack discussion a while ago that having overloads with rvalue ref arguments is too much work and possibly not so useful?